### PR TITLE
Pass the context through to the search service to restore function…

### PR DESCRIPTION
**Pull request by cbeer**
…al compatibility with Blacklight 7.

Some uses of search builders expect access to controller methods (e.g. current user, `can?` from cancan, etc). Instead of making consumers always override the search service, we can provide the context from the controller as a convenience.

### CODE REVIEW

**Comment by jcoyne** [comment]
This seems to totally bust the interface so I'm not keen.  Blacklight-access-controls passes an Ability instance, would that be sufficient?

**Comment by cbeer** [comment]
I don't see how blacklight-access-controls could possibly work right now; as far as I can tell, it's doing the thing that's compatible with Blacklight 6, and that 7 completely removes the ability to do without providing a different SearchService and overriding the `Blacklight::Catalog#search_service` method.

**Comment by jcoyne** [comment]
@cbeer yes, it takes away the ability to configure the search builder....but that aside, could we just pass the Ability as a parameter to a SearchBuilder.  That seems to be the part of the controller that we really need to work with in the SearchBuilder, right?

**Comment by cbeer** [comment]
How are you thinking?

**Comment by jcoyne** [comment]
I'm wondering, that if instead of passing `controller:` in, can we do `ability:` instead?

```ruby
search_service_class.new(config: blacklight_config, user_params: search_state.to_h, ability: current_ability)
```

**Comment by cbeer** [comment]
I guess, but Blacklight doesn't have a dependency on cancan, so it seems a little weird to me to bake that functionality in?

**Comment by jcoyne** [comment]
@cbeer could we encapsulate the fields that get passed to the search builder into a method that can be overridden as needed?

**Comment by cbeer** [comment]
Like bf13e21e?

**Comment by jcoyne** [comment]
Could we do this?
```
search_service_class.new(config: blacklight_config, user_params: search_state.to_h, **search_service_context)
```

File: spec/services/blacklight/search_service_spec.rb [hunk 1]
```diff
@@ -485,4 +486,10 @@
       expect(service.opensearch_response.last).to match_array %w[A B C]
     end
   end
+
+  describe '#context' do
+    it 'has a context attribute' do
+      expect(subject.context).to eq context
```
Comment (jcoyne) [inline spec/services/blacklight/search_service_spec.rb]: Shouldn't `subject.context == { context: context }` ?

Comment (cbeer) [reply spec/services/blacklight/search_service_spec.rb]: I've updated the way it's getting stubbed so this should be fine.

**Comment by jrochkind** [comment]
Not sure if there are docs on how to use this feature somewhere already. So after figuring it out from a code read, I wrote a short (for me haha) blog post. 

https://bibwild.wordpress.com/2019/04/30/blacklight-7-current_user-or-other-request-context-in-searchbuilder-solr-query-builder/

**Pull request merged**